### PR TITLE
`qinfo` module deprecations and removals

### PR DESCRIPTION
**Pull request by isaacdevlugt**
**Context:**

As part of our v0.39 deprecation cycle, `qinfo` will be deprecated. All PRs for deprecating each function can be merged to here,

**Description of the Change:**

Deprecates `qinfo`. Specifically:

- [x] deprecate `qinfo.mutual_info`: https://github.com/PennyLaneAI/pennylane/pull/5917
- [x] deprecate `qinfo.reduced_dm`: https://github.com/PennyLaneAI/pennylane/pull/5915
- [x] deprecate `qinfo.purity`: https://github.com/PennyLaneAI/pennylane/pull/5916
- [x] deprecate `qinfo.vn_entropy`: https://github.com/PennyLaneAI/pennylane/pull/5912
- [x] deprecate `qinfo.vn_entanglement_entropy` https://github.com/PennyLaneAI/pennylane/pull/5914
- [x] deprecate `qinfo.fidelity`: https://github.com/PennyLaneAI/pennylane/pull/5915
- [x] deprecate `qinfo.relative_entropy`: https://github.com/PennyLaneAI/pennylane/pull/5915
- [x] deprecate `qinfo.trace_distance`
- [x] remove `qinfo.classical_fisher`
- [x] remove `qinfo.quantum_fisher`

**Benefits:**

Desired UI and less redundancies.

**Possible Drawbacks:**

None

**Related GitHub Issues:**

[sc-67217] [sc-67216] [sc-66716] [sc-66715] [sc-66714] [sc-66713] [sc-67664] [sc-67665] [sc-67663] [sc-67446]

### CODE REVIEW

**Comment by mudit2812** [comment]
Tests for `qml.vn_entanglement_entropy` are missing, so I'm converting this to a draft once CI finishes.

File: tests/qinfo/test_vn_entanglement_entropy_qinfo.py
Comment (mudit2812) [inline tests/qinfo/test_vn_entanglement_entropy_qinfo.py:1]: File name was changed as the test file for the measurement process has the same name, which pytest does not like.

File: doc/development/deprecations.rst [hunk 1]
```diff
@@ -9,6 +9,11 @@ deprecations are listed below.
 Pending deprecations
 --------------------
 
+* The ``qml.qinfo`` module has been deprecated.
```
Comment (albi3ro) [inline doc/development/deprecations.rst:12]:
```suggestion
* The ``qml.qinfo`` module has been deprecated. Please see the respective functions in `qml.math` or measurements instead.
```

File: doc/code/qml_qinfo.rst [hunk 2]
```diff
@@ -4,6 +4,9 @@ qml.qinfo
 Overview
 --------
 
+.. warning:: 
+    The `qinfo` module is deprecated and scheduled to be removed in v0.40.
```
Comment (albi3ro) [inline doc/code/qml_qinfo.rst:8]: It is potentially worth adding some information about the corresponding functions in `qml.math` and the measurements here to redirect users.

File: pennylane/measurements/vn_entropy.py [hunk 3]
```diff
@@ -61,7 +61,7 @@ def circuit_entropy(x):
 
     .. note::
 
-        Calculating the derivative of :func:`~.vn_entropy` is currently supported when
+        Calculating the derivative of :func:`~pennylane.vn_entropy` is currently supported when
         using the classical backpropagation differentiation method (``diff_method="backprop"``)
         with a compatible device and finite differences (``diff_method="finite-diff"``).
```
Comment (albi3ro) [inline pennylane/measurements/vn_entropy.py:67]: Should we still be pointing to `pennylane.qinfo.transforms.vn_entropy` ?

Comment (mudit2812) [reply pennylane/measurements/vn_entropy.py:67]: It was always pointing to `qml.vn_entropy`, but I had to change this because sphinx was not able to reconcile which target we were referring to.

Comment (albi3ro) [reply pennylane/measurements/vn_entropy.py:67]: Sorry, wasn't able to comment on line 68.  I was speaking about the line just below.

Comment (mudit2812) [reply pennylane/measurements/vn_entropy.py:67]: Good idea, I'll get rid of it here and in other MPs.

File: doc/development/deprecations.rst [hunk 4]
```diff
@@ -22,15 +28,22 @@ Pending deprecations
...
+* `QNode.gradient_fn` is deprecated. Please use `QNode.diff_method` instead. `QNode.get_gradient_fn` can also be used to
+  process the diff method.
+
+  - Deprecated in v0.39
+  - Will be removed in v0.40
+
+* The ``decomp_depth`` argument in ``qml.device`` is deprecated. 
 
-  - Deprecated and Duplicated in v0.38
+  - Deprecated in v0.38
```
Comment (astralcai) [inline doc/development/deprecations.rst:46]: Why are these included in this PR? Is there a poor resolution of merge conflict somewhere?

File: doc/development/deprecations.rst [hunk 5]
```diff
@@ -9,6 +9,12 @@ deprecations are listed below.
 Pending deprecations
 --------------------
 
+* The ``qml.qinfo`` module has been deprecated. Please see the respective functions in the ``qml.math`` and ``qml.measurements``
+  modules instead.
+
+  - Deprecated top level access in v0.39
+  - Top level access removed in v0.40
```
Comment (PietropaoloFrisoni) [inline doc/development/deprecations.rst:16]:
```suggestion
  - Top level access will be removed in v0.40
```

File: doc/development/deprecations.rst (hunk 4 above)
Comment (mudit2812) [reply doc/development/deprecations.rst:46]: Yeah some stuff got duplicated while resolving merge conflicts. Fixed locally now, just waiting for CI to pass before I push the change.

File: doc/development/deprecations.rst [hunk 6]
```diff
@@ -22,15 +28,22 @@ Pending deprecations
...
 
-* The ``decomp_depth`` argument in ``qml.device`` is deprecated. 
+* ``Device``, ``QubitDevice``, and ``QutritDevice`` will no longer be imported top level in v0.40.  They instead
+  we be available as ``qml.devices.LegacyDevice``, ``qml.devices.QubitDevice``, and ``qml.devices.QutritDevice``
+  respectively.
 
-  - Deprecated in v0.38
-  - Will be removed in v0.39
+  - Deprecated top level access in v0.39
+  - Top level access removed in v0.40
```
Comment (PietropaoloFrisoni) [inline doc/development/deprecations.rst:36]: Looks like the following code is a duplicated on the branch:

```
* ``Device``, ``QubitDevice``, and ``QutritDevice`` will no longer be imported top level in v0.40.  They instead
  we be available as ``qml.devices.LegacyDevice``, ``qml.devices.QubitDevice``, and ``qml.devices.QutritDevice``
  respectively.

  - Deprecated top level access in v0.39
  - Top level access removed in v0.40

* `QNode.gradient_fn` is deprecated. Please use `QNode.diff_method` instead. `QNode.get_gradient_fn` can also be used to
  process the diff method.

  - Deprecated in v0.39
  - Will be removed in v0.40
```

File: doc/development/deprecations.rst [hunk 7]
```diff
@@ -9,12 +9,18 @@ deprecations are listed below.
 Pending deprecations
 --------------------
 
+* The ``qml.qinfo`` module has been deprecated. Please see the respective functions in the ``qml.math`` and ``qml.measurements``
+  modules instead.
+
+  - Deprecated top level access in v0.39
+  - Top level access removed in v0.40
```
Comment (PietropaoloFrisoni) [inline doc/development/deprecations.rst:16]:
```suggestion
  - Top level access will be removed in v0.40
```

File: pennylane/devices/default_mixed.py [hunk 8]
```diff
@@ -763,8 +763,9 @@ def execute(self, circuit, **kwargs):
                     self.measured_wires = self.wires
                     return super().execute(circuit, **kwargs)
                 if isinstance(m, (VnEntropyMP, MutualInfoMP)):
-                    # VnEntropy, MutualInfo: Computed for the state prior to measurement. So, readout
-                    # error need not be applied on the corresponding device wires.
+                    # VnEntropy, MutualInfo: Computed for the state
+                    # prior to measurement. So, readout error need not be applied on the
```
Comment (PietropaoloFrisoni) [inline pennylane/devices/default_mixed.py:767]:
```suggestion
                    # prior to measurement. So, readout error need not be applied to the
```

**Review by PietropaoloFrisoni — APPROVED** [review]

**Review by astralcai — APPROVED** [review]
